bin/src/stream_adapter.rs

```diff
@@ -0,0 +1,175 @@
+use async_trait::async_trait;
+
+use fs::cache::tailed_file::LazyLineSerializer;
+use http::types::body::{Line, LineBuilder, LineMeta};
+use http::types::serialize::{
+    IngestLineSerialize, IngestLineSerializeError, SerializeI64, SerializeMap, SerializeStr,
+    SerializeUtf8, SerializeValue,
+};
+use std::collections::HashMap;
+
+pub(crate) enum StrictOrLazyLineBuilder {
+    Strict(LineBuilder),
+    Lazy(LazyLineSerializer),
+}
+
+#[allow(clippy::large_enum_variant)]
+pub(crate) enum StrictOrLazyLines<'a> {
+    Strict(&'a Line),
+    Lazy(LazyLineSerializer),
+}
+
+#[async_trait]
+impl IngestLineSerialize<String, bytes::Bytes, std::collections::HashMap<String, String>>
+    for StrictOrLazyLines<'_>
+{
+    type Ok = ();
+
+    fn has_annotations(&self) -> bool {
```

Comment:
NIT: A macro can be used to avoid repetitive blocks and make sure the calls are the same.

It doesn't have to be part of this PR, maybe a small enhancement in the future.

I've manually checked that the calls are the same per each (except for `has_file()`) 👍

Reply:
Yeh, I didn't want to prematurely optimise with a macro, if we add a third type (which I think we probably will as we add sources or port the existing ones) a macro makes a lot more sense